### PR TITLE
phrase search implemented 

### PR DESCRIPTION
**Pull request by jszymanski97**
phrase search implemented for searching by titles, authors, isbn, min and max price range, and a keyword search in description

### CODE REVIEW

File: src/main/java/mate/project/repository/book/BookSpecificationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    .getSpecification(new BigDecimal[]{searchParameters.minPrice(),
+                            searchParameters.maxPrice()}));
+        }
+        if (searchParameters.descriptionKeyword() != null
+                && searchParameters.descriptionKeyword().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("description")
+                    .getSpecification(searchParameters.descriptionKeyword()));
+        }
+        return spec;
+    }
```
Comment (MateuszRostek) [inline src/main/java/mate/project/repository/book/BookSpecificationBuilder.java]: This method scales very poorly, with each new parameter you have to add another block `if`. Try to modify this method in such a way that with a new filtering parameter you only need to add an item to the map, for example like this:

```java
@Override
    public Specification<Book> build(BookSearchParametersDto searchParametersDto) {
        Specification<Book> specification = Specification.where(null);
        Map<String, String[]> parametersMap = new HashMap<>();
        parametersMap.put("title", searchParametersDto.titles());
        parametersMap.put("author", searchParametersDto.authors());
        parametersMap.put("price",
                new String[]{searchParametersDto.minPrice(), searchParametersDto.maxPrice()});

        for (Map.Entry<String, String[]> entry : parametersMap.entrySet()) {
            String[] paramValues = entry.getValue();
            if (paramValues != null && paramValues.length > 0) {
                specification = specification.and(
                        specificationProviderManager.getSpecificationProvider(entry.getKey())
                                .getSpecification(paramValues));
            }
        }
        return specification;
    }
```

File: src/main/java/mate/project/repository/book/BookSpecificationProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("No specification provider found for key: "
```
Comment (MateuszRostek) [inline src/main/java/mate/project/repository/book/BookSpecificationProviderManager.java]: Create custom exception and use it here.

File: src/main/java/mate/project/service/BookServiceImpl.java [hunk 3]
```diff
@@ -47,4 +51,13 @@ public BookDto updateBookById(Long id, CreateBookRequestDto bookDto) {
...
         bookRepository.deleteById(id);
     }
+
+    @Override
+    public List<BookDto> search(BookSearchParameters searchParameters) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(searchParameters);
+        return bookRepository.findAll(bookSpecification)
+                .stream()
+                .map(bookMapper::toBookDto)
+                .toList();
```
Comment (MateuszRostek) [inline src/main/java/mate/project/service/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAll(bookSpecification).stream()
                .map(bookMapper::toBookDto)
                .toList();
```

**Review by MateuszRostek — CHANGES_REQUESTED** [review]
Nice try! See some advice and implement it.

File: src/main/java/mate/project/repository/book/BookSpecificationBuilder.java (hunk 1 above)
Comment (jszymanski97) [reply src/main/java/mate/project/repository/book/BookSpecificationBuilder.java]: I had issues with converting  the min and max price in to the string parametersMap I tried different ways, code would compile but method wouldnt work properly keep giving me nullPointerExceptions I decided to add it as a sepperate if straight in to spec

**Review by MateuszRostek — APPROVED** [review]
Good job!

**Pull request merged**